src/Game/zFX.cpp

```diff
@@ -1,43 +1,180 @@
+    uint32 randV = xrand();
+    zFX_SpawnBubbleSlam(pos, (randV & 0x1f) + 0x20, lbl_803CD968, lbl_803CD96C, lbl_803CD970);
+}
+
+void zFXPorterWave(const xVec3* pos)
+{
+    xFXRingCreate(pos, &sPorterRing[0]);
+    xFXRingCreate(pos, &sPorterRing[1]);
+}
+
+xFXRing* zFXMuscleArmWave(xVec3* pos)
+{
+    return xFXRingCreate(pos, &sMuscleArmRing);
+}
 
 // func_80092F94
 #pragma GLOBAL_ASM("asm/Game/zFX.s", "zFXGooEnable__FP8RpAtomici")
 
+#if 1
+
 // func_8009337C
 #pragma GLOBAL_ASM("asm/Game/zFX.s", "zFXGoo_SceneEnter__Fv")
 
+#else
+
+// Can't figure out what's going on here.
+void zFXGoo_SceneEnter()
+{
+    for (int32 i = 0; i < 0x18; i++)
+    {
+        zFXGooInstance* goo = &zFXGooInstances[i];
+        memset(goo, 0, 4);
+        goo->state = zFXGooStateInactive;
+    }
+    uint32 gameID = xStrHash("FREEZY_TIMER_TEXTBOX");
+    goo_timer_textbox = (ztextbox*)zSceneFindObject(gameID);
+}
+
+#endif
+
+#if 1
+
 // func_80093404
 #pragma GLOBAL_ASM("asm/Game/zFX.s", "zFXGoo_SceneReset__Fv")
 
+#else
+
+// Idk how to structure the loop properly.
+void zFXGoo_SceneReset()
+{
+    for (int32 i = 3; i != 0; i--)
+    {
+        if ((&zFXGooInstances[(i - 3) * 8 + 0])->state != zFXGooStateInactive)
+        {
+            (&zFXGooInstances[(i - 3) * 8 + 0])->state = zFXGooStateNormal;
+        }
+
+        if ((&zFXGooInstances[(i - 3) * 8 + 1])->state != zFXGooStateInactive)
+        {
+            (&zFXGooInstances[(i - 3) * 8 + 1])->state = zFXGooStateNormal;
+        }
+
+        if ((&zFXGooInstances[(i - 3) * 8 + 2])->state != zFXGooStateInactive)
+        {
+            (&zFXGooInstances[(i - 3) * 8 + 2])->state = zFXGooStateNormal;
+        }
+
+        if ((&zFXGooInstances[(i - 3) * 8 + 3])->state != zFXGooStateInactive)
+        {
+            (&zFXGooInstances[(i - 3) * 8 + 3])->state = zFXGooStateNormal;
+        }
+
+        if ((&zFXGooInstances[(i - 3) * 8 + 4])->state != zFXGooStateInactive)
+        {
+            (&zFXGooInstances[(i - 3) * 8 + 4])->state = zFXGooStateNormal;
+        }
+
+        if ((&zFXGooInstances[(i - 3) * 8 + 5])->state != zFXGooStateInactive)
+        {
+            (&zFXGooInstances[(i - 3) * 8 + 5])->state = zFXGooStateNormal;
+        }
+
+        if ((&zFXGooInstances[(i - 3) * 8 + 6])->state != zFXGooStateInactive)
+        {
+            (&zFXGooInstances[(i - 3) * 8 + 6])->state = zFXGooStateNormal;
+        }
+
+        if ((&zFXGooInstances[(i - 3) * 8 + 7])->state != zFXGooStateInactive)
+        {
+            (&zFXGooInstances[(i - 3) * 8 + 7])->state = zFXGooStateNormal;
+        }
+    }
+}
+
+#endif
+
+#if 1
+
 // func_800934EC
 #pragma GLOBAL_ASM("asm/Game/zFX.s", "zFXGoo_SceneExit__Fv")
 
+#else
+
+// Can't figure out what's going on here.
+void zFXGoo_SceneExit()
+{
+    zFXGooInstance* goo = &zFXGooInstances[0];
+    for (int32 i = 1; i < 0x18; i++)
+    {
+        goo->state = zFXGooStateInactive;
+        memset(goo, 0, 4);
+        goo = &zFXGooInstances[i];
+    }
+}
```

Comment:
This is matching:
```cpp
void zFXGoo_SceneExit()
{
    int32 i;
    zFXGooInstance* goo = zFXGooInstances;
    for (i = 0; i < 0x18; i++)
    {
        memset(goo, 0, 4);
        goo->state = zFXGooStateInactive;
        goo++;
    }
}
```
But for some reason, it seems like the struct is off by 4 bytes. If you add 4 bytes of padding:
```cpp
struct zFXGooInstance
{
    int32 padding;
    // ....
```

I get 🆗

Reply:
If you do add padding to `zFXGooInstance` please make sure to call this out in a comment.